pkg/flow/internal/controller/loader.go

```diff
@@ -607,28 +618,37 @@ func (l *Loader) EvaluateDependencies(updatedNodes []*ComponentNode) []*Componen
 	// During evaluation, if a node's exports change, Flow will add it to updated nodes queue (controller.Queue) and
 	// the Flow controller will call EvaluateDependencies on it again. This results in a concurrent breadth-first
 	// traversal of the nodes that need to be evaluated.
-	toRetry := make([]*ComponentNode, 0)
 	for n, parent := range dependenciesToParentsMap {
 		dependantCtx, span := tracer.Start(spanCtx, "SubmitForEvaluation", trace.WithSpanKind(trace.SpanKindInternal))
 		span.SetAttributes(attribute.String("node_id", n.NodeID()))
 		span.SetAttributes(attribute.String("originator_id", parent.NodeID()))
 
-		// Submit the node for asynchronous evaluation. Don't use range variables in the closure.
-		nodeRef, parentRef := n, parent
-		err := l.workerPool.SubmitWithKey(nodeRef.NodeID(), func() {
-			l.concurrentEvalFn(nodeRef, dependantCtx, tracer, parentRef)
-		})
+		// Submit for asynchronous evaluation with retries and backoff. Don't use range variables in the closure.
+		var (
+			nodeRef, parentRef = n, parent
+			retryBackoff       = backoff.New(ctx, l.backoffConfig)
+			err                error
+		)
+		for retryBackoff.Ongoing() {
+			err = l.workerPool.SubmitWithKey(nodeRef.NodeID(), func() {
+				l.concurrentEvalFn(nodeRef, dependantCtx, tracer, parentRef)
+			})
+			if err != nil {
```

Comment:
Is being overloaded the only reason for an error? What if there is an error which is not due to overloading, for which there is no point in retrying?

Reply:
Yes, being overloaded is the only possible reason for the error here. It will always make sense to retry it, because we don't want to miss evaluating that component, otherwise it will have stale data.